lessons/python_packages/lesson.md

```diff
@@ -0,0 +1,258 @@
+    # long_description=open('README.txt').read(),
+)
+```
+
+## 3. Installing?
+
+Technically, you can install the module by running `python setup.py install`, however it's much more convenient to wrap it all up and distribute online.
+
+### Using git
+
+This is my preferred way of sharing/storing/installing packages. Simply create a git repo of the project directory, and put it somewhere. Then, use `pip` to install it from that repo directly
+
+```
+pip install git+https://github.com/jladan/package_demo.git#egg=measurements
+```
+
+An advantage of this is that you can also install to particular branches, tags, and commits. A common practice is add a tag for the version release `git tag v0.1`. Users of the library can then lock into that particular version,
+
+```
+pip install git+https://github.com/jladan/package_demo.git@v0.1#egg=measurements
+```
+
+
+### Distributing on PyPI
+
+In my opinion, distributing on PyPI is more complicated than with git, and not much more useful.
+
+You'll have to maintain the
+package in two places if you intend on collaborating on development on GitHub
+and distributing through PyPI.
+
+So why distribute on PyPI? If you're proud of your software and want others to use it, this is the best place. It's more visible, easier for your users, and more tightly controlled. The stricter version control (not just whatever commit to whatever repo like git) makes the software more robust and easier to support as a developing team.
+
+For those interested, to upload to PyPI, you'll need
+
+```
+python setup.py sdist
+```
+
+This creates a `dist` directory and adds an archive file of your package.
+
+
+To get set up on PyPI for the first time with a
+new package, [make a new account on PyPI](https://pypi.python.org/pypi), and
+then back in your `package` directory do
+
+```
+python setup.py register
+```
+
+to claim your package name - they must be unique. Then, upload your package; you'll do this every time you want to release a new version of your code:
+
+```
+python setup.py sdist upload
+```
+
+Now, you and anyone else in the world can install your package using the following:
+
+```
+sudo pip install your-package-name
+```
+
+## 4. Document!
+
+We've already begun the documentation by adding a short description, author/maintainer name, and version... but there's a lot more to add. At a minimum, a `README` file should be included, but so should a license text, changes between versions, and actual software documentation. 
+
+### README
+
+A readme file summarizes the software. For Python packages, this can named `README`, `README.rst`, or `README.txt`. The recommendation is to use (reST)[http://docutils.sourceforge.net/rst.html], as this is the standard on PyPI.
+
+```
+===================
+Measurements Module
+===================
+
+**A demo for how to create a python package from existing code**
+
+The actual software provides L1, L2, Supremum, and Manhatten metrics and norms for numpy arrays.
```

Comment:
More significantly, I just realized that I fumbled the name last night. I'll correct this to Hamming in all code and the lesson.